.github/workflows/test.yml

```diff
@@ -4,29 +4,103 @@ on: [push, pull_request]
 
 env:
   CARGO_TERM_COLOR: always
+  # CI builds are closer to a from-scratch build anyways
+  # and incremental compilation hurts reproducibility
+  CARGO_INCREMENTAL: 0
 
 jobs:
-  test:
+  unit-tests:
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
         os: [macos-latest, ubuntu-latest]
     steps:
     - uses: actions/checkout@v2
-    - name: Unit tests
+
+    - uses: Swatinem/rust-cache@v1.3.0
+
+    - name: Build kak-lsp
+      run: cargo test --no-run --locked --release
+
+    - name: Run unit tests
       run: cargo test --release
-    - name: Install system test dependencies
+
+  integration-tests:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [macos-latest, ubuntu-latest]
+    steps:
+    - uses: actions/checkout@v2
+
+    - uses: Swatinem/rust-cache@v1.3.0
+
+    - name: Build kak-lsp
+      run: cargo build --locked --release
+
+    - name: Fetch Kakoune
+      uses: actions/checkout@v2
+      with:
+        repository: 'mawww/kakoune'
+        ref: 'v2021.08.28'
+        # Always relative to `${{ github.workspace }}`
+        path: 'kakoune'
+
+    # Use the current commit hash as part of the cache key, so that we only
+    # need to rebuild when code changes.
+    - name: Get Kakoune commit id
+      id: kakoune-commit-id
+      run: echo "::set-output name=id::$(cd $GITHUB_WORKSPACE/kakoune && git rev-parse HEAD)"
+
+    - name: Cache Kakoune
+      id: cache-kakoune
+      uses: actions/cache@v2
+      with:
+        key: ${{ runner.os }}-kakoune-${{ steps.kakoune-commit-id.outputs.id }}
+        path: |
+          ${{ github.workspace }}/kakoune/src/kak
+          ${{ github.workspace }}/kakoune/src/kak.opt
+          ${{ github.workspace }}/kakoune/src/.*.opt.d
+          ${{ github.workspace }}/kakoune/src/.*.opt.o
+
+    - name: Build Kakoune
+      if: steps.cache-kakoune.outputs.cache-hit != 'true'
+      working-directory: ${{ github.workspace }}/kakoune/src
       run: |
-        (cd && git clone https://github.com/mawww/kakoune)
-        (cd ~/kakoune && git rev-parse HEAD && make && sudo make install)
+        if [ "${{ matrix.os }}" = "macos-latest" ]; then
+          n=$(sysctl -n hw.logicalcpu)
+        else
+          n=$(nproc)
+        fi
+
+        make -j$n
+
+    - name: Touch build targets
+      if: steps.cache-kakoune.outputs.cache-hit
```

Comment:
I think this condition is always true since the variable is just a javascript string.
It hints at why this is needed but it's also confusing. We could move this to the install step, like `make -t && sudo make install`.
However, it seems even cleaner to move this to the build step above:

```
if ${{ steps.cache-kakoune.outputs.cache-hit  }}; then
	make -t
else
	make -j$n
fi
```